client/transaction-pool/src/enactment_state.rs

```diff
@@ -68,32 +78,69 @@ where
 		self.recent_finalized_block
 	}
 
+	/// Returns the recently finalized block.
+	pub fn recent_best_block(&self) -> Block::Hash {
+		self.recent_best_block
+	}
+
 	/// Updates the state according to the given `ChainEvent`, returning
 	/// `Some(tree_route)` with a tree route including the blocks that need to
 	/// be enacted/retracted. If no enactment is needed then `None` is returned.
-	pub fn update<F>(
+	pub fn update<TreeRouteF, IsMajorSyncF, BlockNumberF>(
 		&mut self,
 		event: &ChainEvent<Block>,
-		tree_route: &F,
-	) -> Result<Option<TreeRoute<Block>>, String>
+		tree_route: &TreeRouteF,
+		hash_to_number: &BlockNumberF,
+		is_major_syncing: &IsMajorSyncF,
+	) -> Result<EnactmentAction<Block>, String>
 	where
-		F: Fn(Block::Hash, Block::Hash) -> Result<TreeRoute<Block>, String>,
+		TreeRouteF: Fn(Block::Hash, Block::Hash) -> Result<TreeRoute<Block>, String>,
+		BlockNumberF: Fn(Block::Hash) -> Result<Option<NumberFor<Block>>, String>,
+		IsMajorSyncF: Fn() -> bool,
 	{
+		if is_major_syncing() {
+			log::info!(target: "txpool", "skip maintain: major syncing");
+			self.force_update(event);
+			return Ok(EnactmentAction::Skip)
+		}
+
 		let (new_hash, finalized) = match event {
 			ChainEvent::NewBestBlock { hash, .. } => (*hash, false),
 			ChainEvent::Finalized { hash, .. } => (*hash, true),
 		};
 
+		// do not proceed with txpool maintain if block distance is to high
+		let skip_maintanance =
+			match (hash_to_number(new_hash), hash_to_number(self.recent_best_block())) {
```

Comment:
We should use here `recent_best_block` when `finalized == false` and vice versa with `recent_finalized_block`.

Reply:
Having only single-field updated in `force_update`, yes. Fixed in: https://github.com/paritytech/substrate/pull/13004/commits/c0f8d0fe4f5984a3ac50015dab89c02ec02e4656